### PR TITLE
[YUNIKORN-45] Task is not properly released when they are removed from pending state

### PR DESCRIPTION
**Pull request by yangwwei**
See jira for more info

### CODE REVIEW

**Comment by yangwwei** [comment]
Unit tests included, @wilfred-s , please help to review this. Thanks

File: pkg/cache/task_test.go [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+	app := NewApplication("app01", "root.default",
+		"bob", map[string]string{}, mockedSchedulerApi)
+	task := NewTask("task01", app, mockedContext, pod)
+	assert.Equal(t, task.GetTaskState(), events.States().Task.New)
+
+	// new task
+	event0 := NewSimpleTaskEvent(task.applicationID, task.taskID, events.InitTask)
+	if err := task.handle(event0); err != nil {
+		t.Fatal(err)
+	}
```
Comment (wilfred-s) [inline pkg/cache/task_test.go]: please use `assert.NilError` make the code easier to read

File: pkg/cache/task_test.go [hunk 2]
```diff
@@ -0,0 +1,301 @@
...
+import (
+	"testing"
+
+	"github.com/apache/incubator-yunikorn-k8shim/pkg/client"
+	"github.com/apache/incubator-yunikorn-k8shim/pkg/common/events"
+	"github.com/apache/incubator-yunikorn-scheduler-interface/lib/go/si"
+	"gotest.tools/assert"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	apis "k8s.io/apimachinery/pkg/apis/meta/v1"
```
Comment (wilfred-s) [inline pkg/cache/task_test.go]: please fix grouping: golang, other followed by project

File: pkg/cache/task_test.go [hunk 3]
```diff
@@ -0,0 +1,301 @@
...
+}
+
+func TestReleaseTaskAllocation(t *testing.T) {
+	mockedSchedulerApi := newMockSchedulerAPI()
+	mockedContext := initContextForTest()
+	apiProvider := mockedContext.apiProvider
+	mockedApiProvider, ok := apiProvider.(*client.MockedAPIProvider)
+	if !ok {
+		t.Fatal("expecting MockedAPIProvider")
+	}
```
Comment (wilfred-s) [inline pkg/cache/task_test.go]: we can use `assert.Assert(t, ok, "expecting MockedAPIProvider")`

File: pkg/client/apifactory_mock.go [hunk 4]
```diff
@@ -19,6 +19,7 @@
 package client
 
 import (
+	"github.com/apache/incubator-yunikorn-scheduler-interface/lib/go/si"
```
Comment (wilfred-s) [inline pkg/client/apifactory_mock.go]: import ordering check

File: pkg/common/resource_test.go [hunk 5]
```diff
@@ -243,3 +243,27 @@ func TestNodeResource(t *testing.T) {
...
+	request := CreateReleaseAskRequestForTask("app01", "task01", "default")
+	assert.Assert(t, request.Releases != nil)
+	assert.Assert(t, request.Releases.AllocationsToRelease == nil)
+	assert.Assert(t, request.Releases.AllocationAsksToRelease != nil)
+	assert.Equal(t, len(request.Releases.AllocationsToRelease), 0)
+	assert.Equal(t, len(request.Releases.AllocationAsksToRelease), 1)
+	assert.Equal(t, request.Releases.AllocationAsksToRelease[0].ApplicationID, "app01")
+	assert.Equal(t, request.Releases.AllocationAsksToRelease[0].Allocationkey, "task01")
+	assert.Equal(t, request.Releases.AllocationAsksToRelease[0].PartitionName, "default")
+}
```
Comment (wilfred-s) [inline pkg/common/resource_test.go]: missing newline, please add

File: pkg/cache/task_test.go (hunk 1 above)
Comment (yangwwei) [reply pkg/cache/task_test.go]: Done

File: pkg/cache/task_test.go (hunk 2 above)
Comment (yangwwei) [reply pkg/cache/task_test.go]: done

File: pkg/cache/task_test.go (hunk 3 above)
Comment (yangwwei) [reply pkg/cache/task_test.go]: done

File: pkg/client/apifactory_mock.go (hunk 4 above)
Comment (yangwwei) [reply pkg/client/apifactory_mock.go]: updated

**Comment by wilfred-s** [comment]
committed changes closing PR

**Pull request closed without merging**